lib/active_merchant/billing/gateways/barclaycard_smartpay.rb

```diff
@@ -230,17 +223,41 @@ def build_url(action)
         end
       end
 
-      def address_hash(address)
-        full_address = "#{address[:address1]} #{address[:address2]}" if address
-        street = address[:street] if address[:street]
-        house = address[:houseNumberOrName] ? address[:houseNumberOrName] : full_address.split(/\s+/).keep_if { |x| x =~ /\d/ }.join(' ')
+      def billing_address_hash(options)
+        address = options[:address] || options[:billing_address] if options[:address] || options[:billing_address]
+        street = options[:street] || parse_street(address)
+        house = options[:house_number] || parse_house_number(address)
+
+        create_address_hash(address, house, street)
+      end
+
+      def shipping_address_hash(options)
+        address = options[:shipping_address]
+        street = options[:shipping_street] || parse_street(address)
+        house = options[:shipping_house_number] || parse_house_number(address)
 
+        create_address_hash(address, house, street)
+      end
+
+      def parse_street(address)
+        address_to_parse = "#{address[:address1]} #{address[:address2]}"
+        street = address[:street] || address_to_parse.split(/\s+/).keep_if { |x| x !~ /\d/ }.join(' ')
+        street.empty? ? "Not Provided" : street
+      end
+
+      def parse_house_number(address)
+        address_to_parse = "#{address[:address1]} #{address[:address2]}"
+        house = address[:houseNumberOrName] || address_to_parse.split(/\s+/).keep_if { |x| x =~ /\d/ }.join(' ')
```

Comment:
I know it was assumed in (my) current code, but I don't think there'd ever be an `address[:houseNumberOrName]` or `address[:street]` field sent to the adapter, so I think they can be removed. I think I was anticipating needing this split (which this commit is taking care of), but it wouldn't be sent in the options' `address` element.

Reply:
Though that begs the question if we need top-level street and house_number fields for both billing and shipping address...

Reply:
Agreed about `address[:houseNumberOrName]` or `address[:street]` being unlikely, but will leave in case they are currently being used somewhere since they were pre-existing. 

This does include top-level street and house_number fields for both billing and shipping addresses.